pkg/controller/awselb/aws_elb_controller.go

```diff
@@ -0,0 +1,313 @@
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		c.logger.Infof("Error syncing machine %v: %v", key, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	c.logger.Infof("Dropping machine %q out of the queue: %v", key, err)
+	c.queue.Forget(key)
+}
+
+// syncMachine will sync the machine with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncMachine(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing machine")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing machine")
+	}()
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	machine, err := c.machinesLister.Machines(ns).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Info("machine has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	mLog := clustoplog.WithMachine(c.logger, machine)
+
+	clusterID, ok := machine.Labels[clustopv1.ClusterNameLabel]
+	if !ok {
+		return fmt.Errorf("unable to lookup cluster for machine: %s", machine.Name)
+	}
+
+	// Deep-copy otherwise we are mutating our cache.
+	//m := machine.DeepCopy()
+	coMachineSetSpec, err := controller.MachineSetSpecFromClusterAPIMachineSpec(&machine.Spec)
+	if err != nil {
+		return err
+	}
+
+	region := coMachineSetSpec.ClusterHardware.AWS.Region
+	mLog.Debugf("Obtaining AWS clients for region %q", region)
+	ec2Client, elbClient, err := clustopaws.CreateAWSClients(c.kubeClient, coMachineSetSpec, machine.Namespace, region)
+	if err != nil {
+		return err
+	}
+
+	instance, exists, err := clustopaws.GetInstance(machine, ec2Client)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("no instance found for machine: %s", machine.Name)
+	}
+	mLog = mLog.WithField("instanceID", *instance.InstanceId)
+
+	err = c.addInstanceToELB(instance, fmt.Sprintf("%s-master-external", clusterID), elbClient, mLog)
+	if err != nil {
+		return err
+	}
+	err = c.addInstanceToELB(instance, fmt.Sprintf("%s-master-internal", clusterID), elbClient, mLog)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Controller) addInstanceToELB(
+	instance *ec2.Instance,
+	elbName string,
+	elbClient elbiface.ELBAPI,
+	mLog log.FieldLogger) error {
+
+	registerInput := elb.RegisterInstancesWithLoadBalancerInput{
+		Instances:        []*elb.Instance{{InstanceId: instance.InstanceId}},
+		LoadBalancerName: aws.String(elbName),
+	}
+
+	// This API call appears to be idempotent, so for now no need to check if the instance is
+	// registered first, we can just request that it be added.
+	_, err := elbClient.RegisterInstancesWithLoadBalancer(&registerInput)
+	if err != nil {
+		return err
```

Comment:
It would be nice to have status in the machine that states that the elb registration failed, with a reason why.

Reply:
Condition I assume you meant? (no condition arrays on any of these types yet) Or explicit fields on our provider status?

capi.MachineStatus has an error message/reason field but it's documented as something that should not be used for transient errors.

Reply:
Yes, conditions would be the approach that I would be the most comfortable with. But however it is done, I would like for a user to be able to see that the master machine has been added to the elb or why it hasn't.

Reply:
By the way, this can certainly be done as later work, too.